doc/specs/stdlib_str2num.md

```diff
@@ -0,0 +1,105 @@
+---
+title: str2num
+---
+
+# The `stdlib_str2num` module
+
+This module proposes a function-style interface for string-to-number conversion. It also profits from Fortran's interfaces to implement precision-dependant algorithms to maximize runtime efficiency.
+
+[TOC]
+
+## `to_num` - conversion of strings to numbers
+
+### Status
+
+Experimental
+
+### Description
+
+Convert a string or an array of strings to numerical types.
+
+### Syntax
+
+`number = [[stdlib_str2num(module):to_num(interface)]](string, mold)`
+
+### Arguments
+
+`string`: argument has `intent(in)` and is of type `character(*)`.
+
+`mold`: argument has `intent(in)` and is of numerical type. currently: `integer`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.
+
+### Return value
+
+Return a scalar of numerical type (`integer`, `real32` or `real64`).
+
+### Example
+
+```fortran
+program example_string_to_number
+  use stdlib_kinds, only: dp
+  use stdlib_str2num
+  implicit none
+  character(:), allocatable :: txt
+  real(dp) :: x
+
+  txt = ' 8.8541878128e−12 '
+  x = to_num( txt , x )
+end program example_random_seed
+```
+
+## `to_num_p` - conversion of a stream of values in a string to numbers
+
+### Status
+
+Experimental
+
+### Description
+
+Convert a stream of values in a string to an array of values.
+
+### Syntax
+
+`number = [[stdlib_str2num(module):to_num_p(interface)]](string, mold)`
```

Comment:
Suggestion:

transform it to a subrouine and hide the pointer and the loop inside the subroutine:
```fortran
call to_num_p(string, array)
...
subroutine to_num_p(string, array)
character(*), intent(in), target :: string
real(dp), intent(out) :: array(:)
....
cptr => string

do i = 1, size(array)
...
enddo

end subroutine
```

Reply:
I've been giving some thought to your idea, on one side I think it does facilitate things, but at the same time I feel that it would introduce some particular cases to handle (multiple dimension arrays, stopping on some condition other than the end of the string, etc) providing the low level interface gives the user the freedom to adapt on top of it. What do you think?

Reply:
Yes, it makes sense to provide the low-level interface only. Keep it as is for now